src/KDSoapClient/KDSoapClientInterface.cpp

```diff
@@ -142,12 +144,23 @@ QNetworkRequest KDSoapClientInterfacePrivate::prepareRequest(const QString &meth
     return request;
 }
 
-QBuffer *KDSoapClientInterfacePrivate::prepareRequestBuffer(const QString &method, const KDSoapMessage &message, const KDSoapHeaders &headers)
+QBuffer *KDSoapClientInterfacePrivate::prepareRequestBuffer(const QString &method, const KDSoapMessage &message, const QString &soapAction, const KDSoapHeaders &headers)
 {
     KDSoapMessageWriter msgWriter;
     msgWriter.setMessageNamespace(m_messageNamespace);
     msgWriter.setVersion(m_version);
-    const QByteArray data = msgWriter.messageToXml(message, (m_style == KDSoapClientInterface::RPCStyle) ? method : QString(), headers, m_persistentHeaders, m_authentication);
+
+    KDSoapMessage msg = message;
+    if (m_useWSAddressing) {
+        KDSoapMessageAddressingProperties prop = message.messageAddressingProperties();
+        prop.setAction(soapAction);
```

Comment:
Until now it was up to the caller to configure the addressing properties correctly. Now it's done magically in here, but do we really want that? This would overwrite any action set by the caller.

In other words this patch combines two features:
 - if you don't want the action http parameter, call setUseWSAddressing(true)
 - oh but when you do that, we'll overwrite your earlier call to messageAddressingProperties().setAction(), if any.

It seems more flexible to me to separate both things, especially if, as this patch shows, we end up needing more flexibility than reading the spec initially indicated.

I admit that I don't know how all this interacts with kdwsdl2cpp though, but whatever we decide, we can always adapt the code generator accordingly, if needed.

Reply:
Yeah, it probably makes sense to split the two up. And looking at some of the code generated by kdwsdl2cpp, it could make sense to let it generate the code for writing the action in the KDSoapMessageAddressingProperties. It already generates the KDSoapMessage and has the action string available.

Reply:
My wsdl file contains a  <wsaw:UsingAddressing/> which (I think) refers to the usage of the WS-Addressing:

```
<?xml version="1.0" encoding="utf-8"?>
<wsdl:definitions xmlns:wsdl="http://schemas.xmlsoap.org/wsdl/" xmlns:wsx="http://schemas.xmlsoap.org/ws/2004/09/mex" xmlns:i0="http://localhost/nd/communication/" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" xmlns:wsa10="http://www.w3.org/2005/08/addressing" xmlns:wsp="http://schemas.xmlsoap.org/ws/2004/09/policy" xmlns:wsap="http://schemas.xmlsoap.org/ws/2004/08/addressing/policy" xmlns:msc="http://schemas.microsoft.com/ws/2005/12/wsdl/contract" xmlns:soap12="http://schemas.xmlsoap.org/wsdl/soap12/" xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing" xmlns:wsam="http://www.w3.org/2007/05/addressing/metadata" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:tns="http://tempuri.org/" xmlns:soap="http://schemas.xmlsoap.org/wsdl/soap/" xmlns:wsaw="http://www.w3.org/2006/05/addressing/wsdl" xmlns:soapenc="http://schemas.xmlsoap.org/soap/encoding/" name="MasterData" targetNamespace="http://tempuri.org/">
  <wsp:Policy wsu:Id="WSHttpBinding_IMasterData_policy">
    <wsp:ExactlyOne>
      <wsp:All>
        <sp:TransportBinding xmlns:sp="http://schemas.xmlsoap.org/ws/2005/07/securitypolicy">
          <wsp:Policy>
            <sp:TransportToken>
              <wsp:Policy>
                <sp:HttpsToken RequireClientCertificate="true"/>
              </wsp:Policy>
            </sp:TransportToken>
            <sp:AlgorithmSuite>
              <wsp:Policy>
                <sp:Basic256/>
              </wsp:Policy>
            </sp:AlgorithmSuite>
            <sp:Layout>
              <wsp:Policy>
                <sp:Strict/>
              </wsp:Policy>
            </sp:Layout>
          </wsp:Policy>
        </sp:TransportBinding>
        <wsaw:UsingAddressing/>
```
I agree with both of you about separating the two things.